File: app/src/main/java/it/amonshore/comikkua/data/web/CmkWebViewModel.java
package it.amonshore.comikkua.data.web;

import android.app.Application;
import android.text.TextUtils;

import androidx.annotation.NonNull;
import androidx.lifecycle.AndroidViewModel;
import androidx.lifecycle.LiveData;
import androidx.lifecycle.MutableLiveData;
import androidx.lifecycle.Transformations;
import androidx.lifecycle.ViewModelKt;
import androidx.paging.Pager;
import androidx.paging.PagingConfig;
import androidx.paging.PagingData;
import androidx.paging.PagingLiveData;
import it.amonshore.comikkua.ICallback;
import it.amonshore.comikkua.data.CustomData;
import it.amonshore.comikkua.data.Resource;
import kotlinx.coroutines.CoroutineScope;

public class CmkWebViewModel extends AndroidViewModel {

    private final CmkWebRepository mCmkWebRepository;
    private final MutableLiveData<String> mFilterComics;
    private final LiveData<PagingData<AvailableComics>> mAllAvailableComics;
    private final LiveData<PagingData<AvailableComics>> mFilteredAvailableComics;
    private String mLastFilter;

    public CmkWebViewModel(@NonNull Application application) {
        super(application);

        mCmkWebRepository = new CmkWebRepository(application);
        mFilterComics = new MutableLiveData<>();

        // preparo paging livedata per i comics
        final PagingConfig pagingConfig = new PagingConfig(10, 10, true);
        final CoroutineScope viewModelScope = ViewModelKt.getViewModelScope(this);
        final Pager<Integer, AvailableComics> allComicsPager = new Pager<>(pagingConfig,
                mCmkWebRepository::getAvailableComicsPagingSource);
        mAllAvailableComics = PagingLiveData.cachedIn(PagingLiveData.getLiveData(allComicsPager), viewModelScope);

        mFilteredAvailableComics = Transformations.switchMap(mFilterComics, input -> {
            if (TextUtils.isEmpty(input) || input.equals("%%")) {
                return mAllAvailableComics;
            } else {
                // TODO: deve esserci un modo per farla più performante
                final Pager<Integer, AvailableComics> filteredComicsPager = new Pager<>(pagingConfig,
                        () -> mCmkWebRepository.getAvailableComicsPagingSource(input));
                return PagingLiveData.cachedIn(PagingLiveData.getLiveData(filteredComicsPager), viewModelScope);
            }
        });
    }

    public void setFilter(String filter) {
        mLastFilter = filter;
        mFilterComics.postValue(filter);
    }

    public String getLastFilter() {
        return mLastFilter;
    }

    public LiveData<PagingData<AvailableComics>> getAvailableComics() {
        return mAllAvailableComics;
    }

    public LiveData<PagingData<AvailableComics>> getFilteredAvailableComics() {
        return mFilteredAvailableComics;
    }
}

File: app/src/main/java/it/amonshore/comikkua/data/web/CmkWebViewModel.kt
package it.amonshore.comikkua.data.web

import android.app.Application
import androidx.lifecycle.*
import it.amonshore.comikkua.LogHelper
import it.amonshore.comikkua.containsAll
import it.amonshore.comikkua.splitToWords
import kotlinx.coroutines.FlowPreview
import kotlinx.coroutines.Job
import kotlinx.coroutines.flow.*
import kotlinx.coroutines.launch

private const val FILTER_DEBOUNCE = 300L;

class CmkWebViewModel(application: Application) : AndroidViewModel(application) {

    private val _repository = CmkWebRepositoryKt(application)
    private val _filter = MutableLiveData<String>()

    private var _lastFilter: String = ""
    private var _filteringJob: Job? = null

    @FlowPreview
    private val _filterFlow: Flow<List<String>> = flow {
        // ogni volta che viene modificato il filtro emetto una lista con tutte le parole contenute nel filtro
        // se il filtro è vuoto emetto null
        _filter.asFlow()
            .debounce(FILTER_DEBOUNCE)
            .map { it.trim() }
            .distinctUntilChanged()
            .map {
                if (it.isEmpty()) {
                    emptyList()
                } else {
                    it.splitToWords()
                }
            }
            .collect { emit(it) }
    }

    /**
     * Filtro applicato ai comics letti con [getFilteredAvailableComics]
     * su [AvailableComics.name] e [AvailableComics.publisher]
     */
    var filter: String
        get() = _lastFilter
        set(value) {
            _lastFilter = value
            _filter.postValue(_lastFilter)
        }

    fun useLastFilter(): String {
        _filter.postValue(_lastFilter)
        return _lastFilter
    }

//    /**
//     * Ritorna un [PagingData] con i comics disponibii,
//     * eventualmente filtrati grazie a [CmkWebViewModelKt.filter]
//     */
//    fun getAvailableComics(): LiveData<PagingData<AvailableComics>> {
//        val currentData = _filteredAvailableComics
//        return if (currentData != null) {
//            currentData
//        } else {
//
//            // TODO: An instance of PagingSource was re-used when Pager expected to create a new
//
//            val data = Transformations.switchMap(_filter) {
//                // se il filtro non è nullo sostituisco gli spazi vuoti con "%" e racchiudo tutto tra "%"
//                val likeName = it?.let {
//                    "%${it.replace("\\s+".toRegex(), "%")}%"
//                }
//                _repository.getAvailableComicsPagingSourceLiveData(likeName)
//                        .cachedIn(viewModelScope)
//            }
//            _filteredAvailableComics = data
//            data
//        }
//    }

    /**
     * LiveData con i comics disponibili filtrati grazie alla proprietà [CmkWebViewModel.filter].
     * Se il filtro è vuoto viene emessa una lista vuota.
     */
    @OptIn(FlowPreview::class)
    fun getFilteredAvailableComics(): LiveData<List<AvailableComics>> = liveData {
        _filterFlow
            .onStart { emit(emptyList()) }
            .collectLatest { filter ->
                LogHelper.d("Filtering available comics by $filter")
                // eventuali precedenti operazioni di filtering vengono annullate
                _filteringJob?.cancel();
                // eseguo il filtering in una coroutine in modo da poter annullare la,
                // e con lei anche il FLow creato al suo interno
                _filteringJob = viewModelScope.launch {
                    if (filter.isEmpty()) {
                        _repository.getAvailableComicsFlow()
                            .catch { LogHelper.e("Error reading available comics", it) }
                            .collectLatest { emit(it) }
                    } else {
                        _repository.getAvailableComicsFlow()
                            .map { data ->
                                data.filter { comics ->
                                    comics.searchableName.containsAll(
                                        filter
                                    )
                                }
                            }
                            .catch { LogHelper.e("Error filtering available comics", it) }
                            .collectLatest { emit(it) }
                    }
                }
            }
    }
}

